src/test/java/io/tiledb/spark/NullableAttributesTest.java

```diff
@@ -0,0 +1,518 @@
+
+    // Create and set getDomain
+    Domain domain = new Domain(ctx);
+    domain.addDimension(rows);
+
+    Attribute a1 = new Attribute(ctx, "a1", Datatype.TILEDB_CHAR);
+    a1.setCellValNum(TILEDB_VAR_NUM);
+
+    a1.setNullable(nullable);
+
+    ArraySchema schema = new ArraySchema(ctx, TILEDB_DENSE);
+    schema.setTileOrder(TILEDB_ROW_MAJOR);
+    schema.setCellOrder(TILEDB_ROW_MAJOR);
+    schema.setDomain(domain);
+    schema.addAttribute(a1);
+
+    Array.create(variableAttURI, schema);
+  }
+
+  public void denseArrayVarAttWrite() throws Exception {
+
+    NativeArray a1_data = new NativeArray(ctx, "aabbccddeeffgghh", Datatype.TILEDB_CHAR);
+    NativeArray a1_off =
+        new NativeArray(ctx, new long[] {0, 3, 4, 6, 7, 10, 13, 14}, Datatype.TILEDB_UINT64);
+
+    // byte vector for null values
+    NativeArray a1ByteMap =
+        new NativeArray(ctx, new short[] {1, 0, 1, 0, 1, 1, 0, 1}, Datatype.TILEDB_UINT8);
+
+    // Create query
+    try (Array array = new Array(ctx, variableAttURI, TILEDB_WRITE);
+        Query query = new Query(array)) {
+      query.setLayout(TILEDB_ROW_MAJOR);
+      query.setBufferNullable("a1", a1_off, a1_data, a1ByteMap);
+      // Submit query
+      query.submit();
+    }
+  }
+
+  @Test
+  public void denseArrayVarAttReadTest() throws Exception {
+    denseArrayVarAttCreate(true);
+    denseArrayVarAttWrite();
+    Dataset<Row> dfRead =
+        session().read().format("io.tiledb.spark").option("uri", variableAttURI).load();
+    // dfRead.show();
+    dfRead.createOrReplaceTempView("tmp");
+    List<Row> rows = session().sql("SELECT * FROM tmp").collectAsList();
+    Assert.assertEquals(8, rows.size());
+    // 1st row
+    Row row = rows.get(0);
+    Assert.assertEquals(1, row.getInt(0));
+    Assert.assertEquals("aab", row.getString(1));
+
+    row = rows.get(1);
+    Assert.assertEquals(2, row.getInt(0));
+    Assert.assertNull(row.get(1));
+
+    row = rows.get(2);
+    Assert.assertEquals(3, row.getInt(0));
+    Assert.assertEquals("cc", row.getString(1));
+
+    row = rows.get(3);
+    Assert.assertEquals(4, row.getInt(0));
+    Assert.assertNull(row.get(1));
+
+    row = rows.get(4);
+    Assert.assertEquals(5, row.getInt(0));
+    Assert.assertEquals("dee", row.getString(1));
+
+    row = rows.get(5);
+    Assert.assertEquals(6, row.getInt(0));
+    Assert.assertEquals("ffg", row.getString(1));
+
+    row = rows.get(6);
+    Assert.assertEquals(7, row.getInt(0));
+    Assert.assertNull(row.get(1));
+
+    row = rows.get(7);
+    Assert.assertEquals(8, row.getInt(0));
+    Assert.assertEquals("hh", row.getString(1));
+    return;
+  }
+
+  /*
+  ==============================================================
+                            WRITE TESTS
+  ==============================================================
+   */
+
+  @Test
+  public void denseReadWriteTest() throws Exception {
+    denseArrayCreate(true);
```

Comment:
Let's keep the write tests completely independent from reads. We should create Spark Datasets at runtime and dump them as TileDB arrays, which is also more close to the real scenario. The following plug-and-play snippet could be used as well:

```
public Dataset<Row> createDataset(SparkSession ss) {
    StructField[] structFields =
            new StructField[] {
                    new StructField("rows", DataTypes.IntegerType, false, Metadata.empty()),
                    new StructField("cols", DataTypes.IntegerType, false, Metadata.empty()),
                    new StructField("a1", DataTypes.FloatType, true, Metadata.empty()),
                    new StructField("a2", DataTypes.IntegerType, true, Metadata.empty()),
            };
    List<Row> rows = new ArrayList<>();
    rows.add(RowFactory.create(1, 1, null, 1));
    rows.add(RowFactory.create(1, 2, 3f, 4));
    rows.add(RowFactory.create(2, 1, 4f, null));
    rows.add(RowFactory.create(2, 2, null, 2));
    StructType structType = new StructType(structFields);
    Dataset<Row> df = ss.createDataFrame(rows, structType);
    return df;
  }
```

Reply:
Done